### PR TITLE
Initial DSP-side implementation

### PR DESCRIPTION
**Pull request by contested-space**
This PR adds the basic functionality required for a DSP to decrypt UID2 token.

A `ExUid2.Dsp` server is spawned to periodically query the UID2 operator server to get a fresh keyring, and then stores that keyring in an ETS table. From that point on, tokens can be decrypted via
```
{:ok, %ExUid2.Uid2{uid: the_user_id}} = ExUid2.Dsp.decrypt_token(token)
```

There still needs to be some validating done to the keyring and of course some tests, but I decided to keep those for a separate PR.

### CODE REVIEW

File: lib/ex_uid2/dsp.ex [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  @spec get_keyring() :: Keyring.t()
+  def get_keyring() do
+    [{@keyring, keyring}] = :ets.lookup(@table_name, @keyring)
+    keyring
+  end
+
+  def decrypt_token(token, now_ms \\ :os.system_time(:millisecond)) do
+    keyring = get_keyring()
+    # TODO: validate keys (see encryption.py)
```
Comment (rkallos) [inline lib/ex_uid2/dsp.ex:55]: Will this be part of a separate PR?

File: lib/ex_uid2/dsp.ex [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    :ets.insert(@table_name, {@keyring, fresh_keyring})
+    Process.send_after(self(), :refresh, @refresh_rate_ms)
+    {:noreply, state}
+  end
+
+  @spec get_keyring() :: Keyring.t()
+  def get_keyring() do
+    [{@keyring, keyring}] = :ets.lookup(@table_name, @keyring)
+    keyring
+  end
```
Comment (rkallos) [inline lib/ex_uid2/dsp.ex]: Could this be made private? It's not used outside of this module.

File: lib/ex_uid2/dsp.ex [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  @impl GenServer
+  def init(_opts) do
+    send(self(), :refresh)
+    {:ok, []}
+  end
+
+  @impl GenServer
+  def handle_info(:refresh, state) do
+    fresh_keyring = Api.Sharing.fetch_keyring()
```
Comment (rkallos) [inline lib/ex_uid2/dsp.ex]: You may want to make this robust to crashes, otherwise you may wind up with a crash loop when the service that serves the keyring is unavailable.

File: lib/ex_uid2/dsp.ex [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    {:noreply, state}
+  end
+
+  @spec get_keyring() :: Keyring.t()
+  def get_keyring() do
+    [{@keyring, keyring}] = :ets.lookup(@table_name, @keyring)
+    keyring
+  end
+
+  def decrypt_token(token, now_ms \\ :os.system_time(:millisecond)) do
```
Comment (rkallos) [inline lib/ex_uid2/dsp.ex:54]: A type spec for this function might be nice.

File: lib/ex_uid2/api.ex [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    secret_key = get_secret_key()
+
+    with {:request_body, {:ok, request_body}} <- {:request_body, Request.make_encrypted_request_body(payload, secret_key, ts)},
+         {:response, %Req.Response{status: 200, body: response_body}} <- {:response, post(request_body, path)},
+         {:parsed_response, {:ok, parsed_response}} <- {:parsed_response, Response.decrypt_and_parse(response_body, secret_key)} do
+      {:ok, parsed_response}
+    else
+      {:request_body, error} -> error
+      {:parsed_response, error} -> error
+      {:response, response} -> {:error, response}
```
Comment (rkallos) [inline lib/ex_uid2/api.ex]: It might be nice to point out the non-200 status code in this error tuple.

File: lib/ex_uid2/api.ex [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    end
+  end
+
+  def post(payload, path) do
+    base_url = Application.fetch_env!(:ex_uid2, :base_url)
+    api_key = Application.fetch_env!(:ex_uid2, :api_key)
+
+    Req.post!(base_url <> path, auth: {:bearer, api_key}, body: payload)
```
Comment (rkallos) [inline lib/ex_uid2/api.ex]: Mixing raising exceptions and `{:ok, _} | {:error, _}` can make APIs a bit confusing. I would lean toward using `Req.post/3` instead.

File: lib/ex_uid2/api.ex [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+         {:parsed_response, {:ok, parsed_response}} <- {:parsed_response, Response.decrypt_and_parse(response_body, secret_key)} do
+      {:ok, parsed_response}
+    else
+      {:request_body, error} -> error
+      {:parsed_response, error} -> error
+      {:response, response} -> {:error, response}
+    end
+  end
+
+  def post(payload, path) do
```
Comment (rkallos) [inline lib/ex_uid2/api.ex]: Does this need to be public?

File: lib/ex_uid2/keyring.ex [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    ]
+
+    @spec new(map()) :: ExUid2.Keyring.Key.t()
+    def new(key_map) do
+      %__MODULE__{
+        activates_ms: key_map["activates"],
+        created_ms: key_map["created"],
+        expires_ms: key_map["expires"],
+        id: key_map["id"],
+        secret: key_map["secret"] |> :base64.decode(),
```
Comment (rkallos) [inline lib/ex_uid2/keyring.ex]: `:base64.decode(nil)` will raise an exception. You may want to add a branch to only call `:base64.decode/1` when `key_map["secret"]` is not `nil`

File: lib/ex_uid2/keyring.ex [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+    %__MODULE__{
+      keys: keys,
+      info: Info.new(keyring_map)
+    }
+  end
+
+  @spec get_key(t(), non_neg_integer()) :: {:ok, Key.t()} | {:error, :key_not_found}
+  def get_key(%__MODULE__{} = keyring, id) do
+    case Enum.find(keyring.keys, nil, fn key -> key.id == id end) do
```
Comment (rkallos) [inline lib/ex_uid2/keyring.ex]: Would it be a good idea to store keys in a map keyed by `key.id`?

File: test/ex_uid2_test.exs [hunk 10]
```diff
@@ -0,0 +1,8 @@
+defmodule ExUid2Test do
```
Comment (rkallos) [inline test/ex_uid2_test.exs:1]: Are you going to write tests as part of this PR?

**Review by rkallos — COMMENTED** [review]
* Please commit a `mix.lock` file.
* Please commit a `.gitignore` file.

File: lib/ex_uid2/dsp.ex (hunk 3 above)
Comment (contested-space) [reply lib/ex_uid2/dsp.ex]: Indeed.  I was actually implementing that on my other branch but I think it should move here.

File: lib/ex_uid2/dsp.ex (hunk 1 above)
Comment (contested-space) [reply lib/ex_uid2/dsp.ex:55]: Yes. I want to do it for completeness and clarity's sake, but it's not like there is much we can do with invalid keys anyway, it'll just cause decryption errors.

File: lib/ex_uid2/api.ex (hunk 6 above)
Comment (contested-space) [reply lib/ex_uid2/api.ex]: Good point.

File: lib/ex_uid2/keyring.ex (hunk 8 above)
Comment (contested-space) [reply lib/ex_uid2/keyring.ex]: Oh! Good catch.

File: test/ex_uid2_test.exs (hunk 10 above)
Comment (contested-space) [reply test/ex_uid2_test.exs:1]: No. I am working on tests in another branch to keep the size reasonable (but I _am_ writing tests I swear ;) )

File: README.md [hunk 11]
```diff
@@ -1,20 +1,115 @@
...
+  ExUid2.Dsp
+]
+
+Supervisor.start_link(children, ...)
+```
+
+Once started, the application will periodically request a fresh keyring from the configured UID2 operator server. Encrypted tokens
+can then be decrypted.
+
+```elixir
```
Comment (RAttab) [inline README.md:42]: Might be worth showing what the return value looks like if attempting to decrypt before the keys have been fetched. Or how to check if they keys are available.

File: README.md [hunk 12]
```diff
@@ -1,20 +1,115 @@
...
+### `/v2/key/sharing` request
+While the `/v2/key/sharing` endpoint is undocumented, The workflow to query UID2's endpoints is documented, the 
+[Encrypting Requests and Decrypting Responses](https://unifiedid.com/docs/getting-started/gs-encryption-decryption#encryption-and-decryption-code-examples) documentation seem to apply there. It is what 
+this library is doing.
+
+### Encrypted token envelope
+| Offset (bytes) | Size (bytes) | Description |
+| -------------- | ------------ | ----------- |
+| 0              | 1            | UID2 token version |
+| 1              | 4            | Master key ID. Must be int32 big endian. |
```
Comment (RAttab) [inline README.md]: If this is unsigned then `uint32` would be more appropriate.

File: README.md [hunk 13]
```diff
@@ -1,20 +1,115 @@
...
+| 5              | 16           | Master 128-bit initialization vector (IV), which is used to randomize data encryption. |
```
Comment (RAttab) [inline README.md:99]: Pedantry ahead: IVs are not really used to randomize data encryption as you'll typically see the exact same IV for a given encryption key or method. They're more-or-less the equivalent of salts for hash functions and there to prevent lookup table attacks.

File: lib/ex_uid2/dsp.ex [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    keyring = get_keyring()
```
Comment (RAttab) [inline lib/ex_uid2/dsp.ex]: So if no keyring has been fetched yet then this will throw an exception as there will be no entry in the ets table and you'll get a matchspec error. Returning an error would allow the caller to more cleanly react to this scenario.

File: lib/ex_uid2/api/request.ex [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+      {_, error} -> error
+    end
+  end
+
+  def make_unencrypted_envelope(payload, ts, nonce \\ :crypto.strong_rand_bytes(8))
+  def make_unencrypted_envelope(payload, ts, nonce)
+      when is_binary(payload) and byte_size(nonce) == 8 and is_integer(ts) do
+    ts_bin = :binary.encode_unsigned(ts, :big)
+    pad = 8 - byte_size(ts_bin)
+    encoded_ts = <<0::pad*8, ts_bin::binary>>
```
Comment (RAttab) [inline lib/ex_uid2/api/request.ex:17]: I feel like this is a little sketchy. Do they truly want a 16 byte timestamp? Also `encode_unsigned` outputs the smallest binary representation of the provided integer so if someone provides the ts 2, then you'll only end up with 9 bytes here instead of 16.

File: .formatter.exs [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+    Enum.flat_map(
+      ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}"],
+      &Path.wildcard(&1, match_dot: true)
+    ) --
+      [
+        "lib/ex_uid2/encryption.ex",
+        "lib/ex_uid2/api.ex",
+        "lib/ex_uid2/api/request.ex",
+        "lib/ex_uid2/api/response.ex"
+      ]
```
Comment (RAttab) [inline .formatter.exs:12]: I think that if you were simply increase the column limit that the formatter will allow, you're problems would go away.

Your longest line is about 145 so 150 would do the trick nicely.

File: lib/ex_uid2/api.ex [hunk 17]
```diff
@@ -0,0 +1,29 @@
+defmodule ExUid2.Api do
+  alias ExUid2.Api.Request
+  alias ExUid2.Api.Response
+
+  def post_encrypted_request(path, payload, ts \\ :os.system_time(:millisecond)) do
+    secret_key = get_secret_key()
+
+    with {:request_body, {:ok, request_body}} <- {:request_body, Request.make_encrypted_request_body(payload, secret_key, ts)},
```
Comment (RAttab) [inline lib/ex_uid2/api.ex:9]: For my own education, why not simply do the following:
```suggestion
    with {:ok, request_body} <- Request.make_encrypted_request_body(payload, secret_key, ts),
```
Is the extra atom wrapped in a tuple there for documentation purposes? Note that the atom is generally repeated in the name of the binded values on the left side so the reader can, with a little more effort, fairly easily deduce what's going on.

File: lib/ex_uid2/api/request.ex (hunk 15 above)
Comment (RAttab) [reply lib/ex_uid2/api/request.ex:17]: I completely misread the code...

File: lib/ex_uid2/encryption.ex [hunk 18]
```diff
@@ -0,0 +1,41 @@
+defmodule ExUid2.Encryption do
+  alias ExUid2.Encryption.EncryptedToken
+  alias ExUid2.Encryption.Identity
+  alias ExUid2.Encryption.MasterPayload
+  alias ExUid2.Uid2
+  alias ExUid2.Keyring
+
+  @spec decrypt_v2_token(binary(), Keyring.t(), integer()) :: Uid2.t() | {:error, any()}
+  def decrypt_v2_token(
```
Comment (RAttab) [inline lib/ex_uid2/encryption.ex:9]: Is v2 meant to only decrypt to UID2? Just wondering how do you tell which version of the decryption you're supposed to use.

File: lib/ex_uid2/keyring.ex [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+        master_keyset_id: keyring_map["master_keyset_id"],
+        site_data: keyring_map["site_data"],
+        token_expiry_seconds: keyring_map["token_expiry_seconds"]
+      }
+    end
+  end
+
+  @spec new(map()) :: ExUid2.Keyring.t()
+  def new(%{"keys" => raw_keys} = keyring_map) when not is_nil(raw_keys) do
+    keys = Enum.map(raw_keys, fn raw_key -> Key.new(raw_key) end)
```
Comment (RAttab) [inline lib/ex_uid2/keyring.ex]: Do you happen to know how big the list of keys will typically be? If it gets quite large then it might be worth turning it into a map so that you don't have to linearly traverse the list every time you want to decrypt something.

File: lib/ex_uid2/api.ex (hunk 17 above)
Comment (RAttab) [reply lib/ex_uid2/api.ex:9]: Ah I think I understand. The extra atom is used in the else statement to extract relevant error messages based on which stage failed.

**Review by RAttab — COMMENTED** [review]
Finished a first pass at the code. Let me know if you'd like me to take another read.

File: .formatter.exs (hunk 16 above)
Comment (contested-space) [reply .formatter.exs:12]: One does not simply. There is no option for that in the formatter config, it would require making a plugin which I feel would add noise to this PR.

File: lib/ex_uid2/api.ex (hunk 17 above)
Comment (rkallos) [reply lib/ex_uid2/api.ex:9]: That's exactly right :)

File: README.md (hunk 13 above)
Comment (contested-space) [reply README.md:99]: I used the definition from UID2's doc:
https://unifiedid.com/docs/getting-started/gs-encryption-decryption
> 96-bit initialization vector (IV), which is used to randomize data encryption.

but I can change it.

File: lib/ex_uid2/encryption.ex (hunk 18 above)
Comment (contested-space) [reply lib/ex_uid2/encryption.ex:9]: Yes, V2 is meant to decrypt only UID2. Technically, we can branch on which version to decrypt by looking at the first 2 bytes of the decoded payload.

If the first byte is `2`, then it's a V2 token. Otherwise, if the second byte holds `112`, it's a V3 token. If it holds `128`, it's a V4 token. I don't know how to identify V1 tokens.

In our case, we're currently only interested in decrypting V2 tokens, so a basic check is made [here](https://github.com/market-ops/ex_uid2/blob/dev/lib/ex_uid2/encryption/encrypted_token.ex#L22) and if it's not a V2 token, it will return `{:error, :invalid_v2_token}`. If we ever need to branch on different token versions, it'll be easy to replace the call to `Encryption.decrypt_v2_token/3` with something that parses and branches first, but for now since we only need to support V2 that seemed unnecessary.

File: lib/ex_uid2/dsp.ex (hunk 3 above)
Comment (contested-space) [reply lib/ex_uid2/dsp.ex]: I have added an exponential backoff mechanism when fetching the keyring and made decryption attempts when the keyring is unavailable return an error tuple.

File: lib/ex_uid2/dsp.ex (hunk 14 above)
Comment (contested-space) [reply lib/ex_uid2/dsp.ex]: It now returns an error.

File: lib/ex_uid2/keyring.ex (hunk 19 above)
Comment (contested-space) [reply lib/ex_uid2/keyring.ex]: They are now stored into a map by ID.

File: lib/ex_uid2/api.ex [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+         {:response, {:ok, %Req.Response{status: 200, body: response_body}}} <- {:response, post(request_body, path)},
+         {:parsed_response, {:ok, parsed_response}} <- {:parsed_response, Response.decrypt_and_parse(response_body, secret_key)} do
+      {:ok, parsed_response}
+    else
+      {:request_body, error} -> error
+      {:response,{:ok, %Req.Response{status: _}} = response} -> {:error, response}
+      {:response, error} -> error
+      {:parsed_response, error} -> error
```
Comment (rkallos) [inline lib/ex_uid2/api.ex]: If you match `{:response, {:ok, %Req.Response{status: _}} = response}` as the first head here, you could combine the rest of them into `{_, error} -> error`. I'm not sure if that's a good idea or not, but it's a possibility.

File: lib/ex_uid2/api.ex [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  defp post(payload, path) do
+    base_url = Application.fetch_env!(:ex_uid2, :base_url)
+    api_key = Application.fetch_env!(:ex_uid2, :api_key)
+
+    Req.post(base_url <> path, auth: {:bearer, api_key}, body: payload)
+  end
+
+  def get_secret_key() do
```
Comment (rkallos) [inline lib/ex_uid2/api.ex]: It looks like this could be private, unless you want to expose it.

File: lib/ex_uid2/dsp.ex [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+        nb_attempts = state.fetch_keyring_attempts
+        retry_time_ms = round(:math.pow(2, nb_attempts))
+
+        Logger.warning(
+          "Failed to fetch keyring: #{inspect(error)}. Retrying in #{retry_time_ms} ms."
+        )
+
+        Process.send_after(self(), :refresh, retry_time_ms)
+        # Exponential backoff maxes out at a ~32 seconds interval
+        nb_attempts = min(nb_attempts + 1, 15)
```
Comment (rkallos) [inline lib/ex_uid2/dsp.ex]: For the sake of clarity, I would store the true number of attempts in the state, and instead call `min/2` in the call to `:math.pow(2, ...)`

Comment (rkallos) [reply lib/ex_uid2/dsp.ex]: I also think that the first 8ish retries will be a bit aggressively fast, though I don't really know what to suggest as an alternative.

File: lib/ex_uid2/api.ex (hunk 20 above)
Comment (contested-space) [reply lib/ex_uid2/api.ex]: Oh indeed, we're not doing anything with those atoms anyway.

File: lib/ex_uid2/dsp.ex (hunk 22 above)
Comment (contested-space) [reply lib/ex_uid2/dsp.ex]: I changed it to be a base of 500ms and doubling up to 32s.

File: README.md (hunk 13 above)
Comment (RAttab) [reply README.md:99]: Na it's ok. If that's how they choose to interpret it,better to stay consistent.

File: .formatter.exs (hunk 16 above)
Comment (RAttab) [reply .formatter.exs:12]: Wow ok, first time I see a formatter not letting you tweak things...

File: lib/ex_uid2/encryption.ex (hunk 18 above)
Comment (RAttab) [reply lib/ex_uid2/encryption.ex:9]: Alrighty, thanks for the clarification.

File: lib/ex_uid2/dsp.ex [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+  use GenServer
+
+  @table_name __MODULE__
+  @keyring :uid2_decryption_keyring
+  @refresh_rate_ms 5000
+
+  require Logger
+
+  def start_link(opts \\ []) do
+    :ets.new(@table_name, [:named_table, {:read_concurrency, true}, :public])
```
Comment (rkallos) [inline lib/ex_uid2/dsp.ex]: Is there a particular reason this is in `start_link/1`, giving the table the lifetime of the caller (usually a supervisor), instead of in `init/1`?

Comment (contested-space) [reply lib/ex_uid2/dsp.ex]: Oh good catch, no good reason. I'm moving it to `init/1`.
